src/sinks/loki/event.rs

```diff
@@ -22,8 +28,27 @@ impl Encoder<Vec<LokiRecord>> for LokiBatchEncoder {
         writer: &mut dyn io::Write,
     ) -> io::Result<usize> {
         let batch = LokiBatch::from(input);
-        let body = serde_json::json!({ "streams": [batch] });
-        let body = serde_json::to_vec(&body)?;
+        let body = match self.0 {
+            LokiBatchEncoding::Json => {
+                let body = serde_json::json!({ "streams": [batch] });
+                serde_json::to_vec(&body)?
+            }
+            LokiBatchEncoding::Protobuf => {
+                let labels = batch.stream;
+                let entries = batch
+                    .values
+                    .iter()
+                    .map(|event| {
+                        loki_logproto::util::Entry(
+                            event.timestamp,
+                            String::from_utf8_lossy(&event.event).into_owned(),
+                        )
+                    })
+                    .collect();
+                let batch = loki_logproto::util::Batch(labels, entries);
+                batch.encode()
+            }
+        };
         writer.write_all(&body).map(|()| body.len())
```

Comment:
I think we should be able to make use of the `write_all()` helper I added to sinks/util/encoding.rs .
Similar to how the `amqp sink` is using it here: https://github.com/vectordotdev/vector/blob/d100bf90217e29250fc5c5cedf048fdc6f142a2a/src/sinks/amqp/encoder.rs#L26-L28

Would end up looking something like this:

```suggestion
        write_all(writer, input.len(), &body).map(|()| body.len())
```

Reply:
I suspect the code was written prior to the helper, poor PR has been hanging for a while because of us 😭

Reply:
Replaced in https://github.com/vectordotdev/vector/pull/12927/commits/ac7727b0df287a72c34feb33d6f6a6a322602bbc